### PR TITLE
Sf emails exporter   group emails by case number and save as json in s3

### PR DESCRIPTION
**Pull request by david-pepper**
## What does this change?
Add logic to group the emails from Salesforce by Parent CaseNumber and save as json files to S3 with filename=casenumber

## How to test
run the lambda and verify that json files created in S3. 

Each Json file should:

- contain all emails relating to the Case

- be named by CaseNumber

## Coming Next
When writing to S3, rather than overwrite the file, append to it instead

### Previous PRs

[Create Basic Lambda - Sf emails to s3 exporter #1316](https://github.com/guardian/support-service-lambdas/pull/1316)

[Sf email exporter - authenticate with sf and get emails #1317](https://github.com/guardian/support-service-lambdas/pull/1317)

### CODE REVIEW

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala [hunk 1]
```diff
@@ -1,15 +1,35 @@
...
+    }{
+      emailsForExportFromSf
+        .records
+        .groupBy(_.Parent.CaseNumber)
+        .foreach(p =>
+          writeEmailsJsonToS3(
+            p._1,
+            p._2.asJson.toString()
+          )
+        )
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala:32]:
```suggestion
        .foreach{ case(caseNumber, caseRecords) =>
          writeEmailsJsonToS3(
            caseNumber,
            caseRecords.asJson.toString()
          )
        }
```

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      .key(s"${fileName}.json")
+      .build()
+
+    val requestBody = RequestBody.fromString(caseEmailsJson, StandardCharsets.UTF_8)
+
+    val uploadAttempt = for {
+      result <- s3Write(putRequest, requestBody)
+    } yield {
+      result
+    }
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala:27]:
```suggestion
    val uploadAttempt = s3Write(putRequest, requestBody)
```

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala (hunk 1 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala:45]: Done

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala (hunk 2 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala]: Done

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  def writeEmailsJsonToS3(fileName: String, caseEmailsJson: String): Unit = {
+
+    val putRequest = PutObjectRequest.builder
+      .bucket("emails-from-sf")
+      .key(s"${fileName}.json")
+      .build()
+
+    val requestBody = RequestBody.fromString(caseEmailsJson, StandardCharsets.UTF_8)
+
+    s3Write(putRequest, requestBody)
```
[the remark this answers is not in the document]
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala:23]: Done

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/SalesforceConfig.scala [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
-  apiVersion: String
-)
+                             authUrl: String,
+                             clientId: String,
+                             clientSecret: String,
+                             userName: String,
+                             password: String,
+                             token: String,
+                             apiVersion: String
+                           )
```
[the remark this answers is not in the document]
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/SalesforceConfig.scala:11]: Fixed

**Comment by david-pepper** [comment]
@kelvin-chappell should be ready for your perusal

**Review by kelvin-chappell — APPROVED** [review]
👍

**Pull request merged**